lib/SILOptimizer/Mandatory/OSLogOptimization.cpp

```diff
@@ -942,22 +903,34 @@ static void replaceAllUsesAndFixLifetimes(SILValue foldedVal,
     return;
   }
 
-  // Here, originalVal is not owned. Hence, borrow form foldedVal and use the
-  // borrow in place of originalVal. Also, destroy foldedVal at the end of its
-  // lifetime.
+  // Here, originalVal is guaranteed. It must belong to a borrow scope that
+  // begins at a scope introducing instruction e.g. begin_borrow or load_borrow.
+  // The foldedVal should also have been inserted at that point. Therefore,
+  // create a borrow of foldedVal at the beginning of the scope and use the
+  // borrow in place of the originalVal. Also, end the borrow and destroy
+  // foldedVal at the end of the borrow scope.
   assert(originalVal.getOwnershipKind() == ValueOwnershipKind::Guaranteed);
 
-  SILBuilderWithScope builder(&*std::next(foldedInst->getIterator()));
-  BeginBorrowInst *borrow =
-      builder.createBeginBorrow(foldedInst->getLoc(), foldedVal);
+  BorrowScopeIntroducingValue originalScopeBegin =
+      getUniqueBorrowScopeIntroducingValue(originalVal);
+  SILInstruction *scopeBeginInst =
+      originalScopeBegin.value->getDefiningInstruction();
+  assert(scopeBeginInst);
+
+  SILBuilderWithScope builder(scopeBeginInst);
+  SILValue borrow =
+      builder.emitBeginBorrowOperation(scopeBeginInst->getLoc(), foldedVal);
```

Comment:
Are you just relying on the SIL verifier to catch a foldedVal that does not dominate the original borrow? Again, I'm not sure if this should be a bail out, or explanatory comment.

Reply:
The foldedVal is inserted at the beginning of the borrow scope, so it must dominate the original borrow.  That is the assumption here.